### PR TITLE
Plot nans

### LINKED ISSUE #1780: DataArray.plot raises exception if contents are all NaN
#### Code Sample, a copy-pastable example if possible

```python
xr.DataArray(np.full((2, 2), 0)).plot.imshow()  # works
xr.DataArray(np.full((2, 2), np.nan)).plot.imshow()  # doesn't
```
#### Problem description

If you try to plot a `DataArray` which is entirely filled with `NaN`, you get an exception.  This is *really, really* annoying for people doing satellite image analysis, especially with timeseries where just one step might be very cloudy.

#### Expected Output

Plot of the array extent, entirely in the missing-value colour as for partially-missing data.

#### Output of ``xr.show_versions()``

Confirmed on Windows/Linux/OSX, Xarray 0.9.6 and 0.10.0; sample show_versions below.  It's a pretty obvious error though in failing to handle the no-non-missing-data case when determining data range.

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-696.10.3.el6.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_AU.UTF-8
LOCALE: en_AU.UTF-8

xarray: 0.10.0
pandas: 0.20.3
numpy: 1.13.3
scipy: 0.19.1
netCDF4: 1.3.1
h5netcdf: None
Nio: None
bottleneck: 1.2.1
cyordereddict: None
dask: 0.15.3
matplotlib: 2.1.0
cartopy: None
seaborn: 0.8.0
setuptools: 36.5.0.post20170921
pip: 9.0.1
conda: 4.3.30
pytest: 3.2.1
IPython: 6.1.0
sphinx: 1.6.3
</details>

### PR DESCRIPTION
**Pull request by Zac-HD**
- [x] Closes #1780
 - [x] Tests added (for all bug fixes or enhancements)
 - [x] Tests passed (for all non-documentation changes)
 - [x] Passes ``git diff upstream/master **/*py | flake8 --diff`` (remove if you did not edit any Python files)
 - [x] Fully documented, including `whats-new.rst` for all changes

CC @fmaussion for review; @BexDunn for interest

### CODE REVIEW

File: doc/whats-new.rst [hunk 1]
```diff
@@ -45,7 +45,10 @@ Bug fixes
   coordinates of target, destination and keys. If there are any conflict among
   these coordinates, ``IndexError`` will be raised.
   By `Keisuke Fujii <https://github.com/fujiisoup>`_.
-
+- Bug fixes in :py:meth:`DataArray.plot.imshow`: all-NaN arrays and arrays
+  with size one in some dimension can now be plotted, which is good for
+  exploring satellite imagery.
```
Comment (fmaussion) [inline doc/whats-new.rst]: Add "(:issue:`1780`)"

File: xarray/plot/plot.py [hunk 2]
```diff
@@ -581,8 +580,14 @@ def imshow(x, y, z, ax, **kwargs):
                          'pcolormesh or contour(f)')
 
     # Centering the pixels- Assumes uniform spacing
-    xstep = (x[1] - x[0]) / 2.0
-    ystep = (y[1] - y[0]) / 2.0
+    try:
+        xstep = (x[1] - x[0]) / 2.0
+    except IndexError:
+        xstep = .1
```
Comment (fmaussion) [inline xarray/plot/plot.py]: nit: I find `xstep = .5` more intuitive here

File: xarray/plot/plot.py [hunk 3]
```diff
@@ -581,8 +580,14 @@ def imshow(x, y, z, ax, **kwargs):
...
+    try:
+        ystep = (y[1] - y[0]) / 2.0
+    except IndexError:
+        ystep = .1
```
Comment (fmaussion) [inline xarray/plot/plot.py]: here too

File: xarray/tests/test_plot.py [hunk 4]
```diff
@@ -624,6 +624,13 @@ def test_plot_nans(self):
         clim2 = self.plotfunc(x2).get_clim()
         self.assertEqual(clim1, clim2)
 
+    def test_can_plot_all_nans(self):
+        # regression test for issue #1780
+        DataArray(np.full((2, 2), np.nan)).plot.imshow()
```
Comment (fmaussion) [inline xarray/tests/test_plot.py]: Do this instead:

```python
self.plotfunc(DataArray(np.full((2, 2), np.nan)))
```

`Common2dMixin` will check that all 2d functions work consistently

File: xarray/tests/test_plot.py [hunk 5]
```diff
@@ -624,6 +624,13 @@ def test_plot_nans(self):
...
+
+    def test_can_plot_axis_size_one(self):
+        DataArray(np.ones((1, 1))).plot.imshow()
```
Comment (fmaussion) [inline xarray/tests/test_plot.py]: Here too

File: xarray/plot/plot.py (hunk 2 above)
Comment (fmaussion) [reply xarray/plot/plot.py]: ignore my comment: it doesn't matter

File: doc/whats-new.rst [hunk 6]
```diff
@@ -32,24 +32,32 @@ Enhancements
...
   coordinates of target, destination and keys. If there are any conflict among
   these coordinates, ``IndexError`` will be raised.
   By `Keisuke Fujii <https://github.com/fujiisoup>`_.
 
+- Properly point :py:meth:`DataArray.__dask_scheduler__` to
+  ``dask.threaded.get``.  By `Matthew Rocklin <https://github.com/mrocklin>`_.
+- Bug fixes in :py:meth:`DataArray.plot.imshow`: all-NaN arrays and arrays
+  with size one in some dimension can now be plotted, which is good for
+  exploring satellite imagery.  (:issue:`1780`)
+- Bug fixes in :py:meth:`DataArray.plot.imshow`: all-NaN arrays and arrays
```
Comment (shoyer) [inline doc/whats-new.rst]: This entry is now repeated twice :)

File: xarray/plot/plot.py (hunk 2 above)
Comment (shoyer) [reply xarray/plot/plot.py]: Either way, it would be good to add a comment here noting that the choice is arbitrary

File: xarray/plot/utils.py [hunk 7]
```diff
@@ -165,6 +165,10 @@ def _determine_cmap_params(plot_data, vmin=None, vmax=None, cmap=None,
 
     calc_data = np.ravel(plot_data[~pd.isnull(plot_data)])
 
+    # Handle all-NaN input data gracefully
+    if calc_data.size == 0:
+        calc_data = np.array(0.0)
```
Comment (shoyer) [inline xarray/plot/utils.py]: I suppose the colorbar get the wrong data in this case, but it's still better than an error message?

Again, please comment that the value 0 is arbitrary.

Comment (Zac-HD) [reply xarray/plot/utils.py]: There *is* no data to give the colorbar in this case, so it uses Matplotlib defaults.  Works beautifully for faceted plots of image timeseries though 😄

File: doc/whats-new.rst (hunk 6 above)
Comment (Zac-HD) [reply doc/whats-new.rst]: Ah, rebasing!  Fixed 😄

**Comment by shoyer** [comment]
Thanks @Zac-HD !

**Comment by Zac-HD** [comment]
No problem!  Do you have any idea when this might be released?  I can install from git in my own environment, but I probably can't convince GA to do that in their production environment 😄

**Comment by shoyer** [comment]
There are a few other bugs that fell out of the v0.10 release. Once we fix those it would make sense issue another release soon... Maybe within a few weeks?

**Pull request merged**